File: 백준/Gold/9251. LCS/LCS.cc
#include <cstdio>
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <cstdlib>
#include <algorithm>
#include <map>
#include <string>

using namespace std;
int dp[1001][1001] = { 0, };
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	

	string str1 = {'0'}, str2 = {'0'}, tmp;
	cin >> tmp;
	str1.append(tmp);
	cin >> tmp;
	str2.append(tmp);
	for (int i = 1; i < str2.size(); i++) {
		for (int j = 1; j < str1.size(); j++) {
			if (str2[i] == str1[j]) {
				dp[i][j] = dp[i - 1][j - 1] + 1;
			}
			else {
				dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
			}
		}
	}

	cout << dp[str2.size()-1][str1.size() - 1];
}

File: 백준/Gold/9251. LCS/LCS.py
a = [0] + list(input())
b = [0] + list(input())

dp = [[0 for _ in range(1001)] for _ in range(1001)]
for i in range(1, len(a)):
    for j in range(1, len(b)):
        if a[i] == b[j]:
            dp[i][j] = dp[i - 1][j - 1] + 1
            pass
        else:
            dp[i][j] = max(dp[i][j - 1], dp[i - 1][j])
            pass
print(dp[len(a) - 1][len(b) - 1])
